torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -1823,11 +1826,18 @@ class ShapePropagator {
         node->matches(
             "aten::sub(Tensor self, Tensor other, *, Scalar alpha) -> Tensor") ||
         node->matches("aten::mul(Tensor self, Tensor other) -> Tensor")) {
-      // These nodes and "div" handle tensors of different shapes internally,
-      // so there's no need to insert explicit expand nodes. Note that "div" is
-      // handled by the fallthrough because it's not always safe to run it due
-      // to integer divide-by-zero.
+      // These nodes handle tensors of different shapes internally, so there's
+      // no need to insert explicit expand nodes.
       return PropagateShapeOnNodeByRunningIt(node);
+    } else if (node->matches("aten::div(Tensor self, Tensor other) -> Tensor")) {
+      // "div" handle tensors of different shapes internally, so there's no need
+      // to insert explicit expand nodes.
+      // Note that this function could be merged to the one above , but "div" is
+      // not always safe to run by itself due to integer divide-by-zero.
+      // We fake the execution by running "mul" operation instead.
+      auto op = getOperatorForLiteral(
```

Comment:
Are we sure that `mul` and `div` have the same dtype promotion rules?

Reply:
```
import torch

def t_promo(l_dtype, r_dtype):
    x = torch.ones(5, 5, dtype=l_dtype, device="cuda")
    y = torch.ones(5, 5, dtype=r_dtype, device="cuda")
    out_div = x / y
    out_mul = x * y
    print(out_div.dtype, ", ", out_mul.dtype)
    assert(out_div.dtype == out_mul.dtype)

# no bool tested
dtypes = [torch.float32, torch.float64, torch.float16, torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64]

for l in dtypes:
  for r in dtypes:
    t_promo(l, r)

```

Maybe someone who knows details in type promotion would be able to answer this better.
Empirically they seem to converge and I can't think of a good reason for them to differ.